README.md
# VehicleEditText

[ ![Download](https://api.bintray.com/packages/relish-wang/maven/vehicleedittext/images/download.svg?version=1.0.1) ](https://bintray.com/relish-wang/maven/vehicleedittext/1.0.1/link)

Android用于车牌号输入的自定义键盘.仿"支付宝-车主服务-添加车辆-车牌号"输入框展示的车牌号输入键盘。

以下是本作与支付宝的对比图:
![对比图](./art/compare.png)

## 一、效果展示

![效果展示](./art/view.gif)

## 二、如何使用

在`app/build.gradle`的`dependencies`节点下添加:
```groovy
implementation "wang.relish.widget:vehicleedittext:1.0.2"
```

### 1 使用原生EditText(**推荐**)
如果你不需要为EditText设置以下的监听器:
- View.OnTouchListener
- View.OnFocusChangeListener
- View.OnKeyListener
你可以使用下面这种侵入性小的方式。

```java
EditText vehicleEditText = findViewById(R.id.vet);
VehicleKeyboardHelper.bind(vehicleEditText); // 为输入框绑定车牌号输入键盘
```

### 2 使用wang.relish.widget.VehicleEditText(不推荐)

和正常的EditText一样使用即可。

不过需要注意如果你需要设置以下监听器:
- View.OnTouchListener
- View.OnFocusChangeListener
- View.OnKeyListener

请使用VehicleEditText的
- setOnTouchListener2
- setOnFocusChangeListener2
- setOnKeyListener2

因为这些监听器已经在VehicleKeyboardHelper中被使用了, 重复设置会覆盖原有的功能。

```xml
<wang.relish.widget.VehicleEditText
    android:id="@+id/vet"
    android:layout_width="match_parent"
    android:layout_height="wrap_content" />
```

## 三、注意事项

要求创建的EditText所持有的Context是Activity, 否则无法正常弹出车牌号输入法。// 暂时未修复这个问题(欢迎Merge Request)

## 四、混淆配置

无

## 五、升级日志

[CHANGELOG.md](./CHANGELOG.md)

## 六、致谢

感谢[pxxhbc](https://github.com/pxxhbc)提出此项目的缺陷。

CHANGELOG.md
# CHANGELOG

1.0.2
------
- 修复横竖屏切换的闪退问题
- 修复横屏【删除图标】变形

1.0.1
------
- 修复工程构建问题。
- 提升minSdkVersion至API21(5.0)

1.0.0
------
第一次发布

app/src/main/java/wang/relish/widget/vehicleedittext/sample/TestDialog.java
package wang.relish.widget.vehicleedittext.sample;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.Gravity;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import wang.relish.widget.vehicleedittext.VehicleKeyboardHelper;

/**
 * @author wangxin
 * @since 20190929
 */
public class TestDialog extends Dialog {

    public TestDialog(@NonNull Context context) {
        super(context);
    }

    public TestDialog(@NonNull Context context, int themeResId) {
        super(context, themeResId);
    }

    protected TestDialog(@NonNull Context context, boolean cancelable, @Nullable OnCancelListener cancelListener) {
        super(context, cancelable, cancelListener);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.dialog_test);

        EditText et = findViewById(R.id.et_dialog_input);
        VehicleKeyboardHelper.bind(getOwnerActivity(),et);

        Window window = getWindow();
        if (window != null) {
            WindowManager.LayoutParams params = window.getAttributes();
            params.gravity = Gravity.CENTER | Gravity.CENTER_HORIZONTAL;
            window.setAttributes(params);
        }
    }
}


app/src/main/java/wang/relish/widget/vehicleedittext/sample/MainFragment.java
package wang.relish.widget.vehicleedittext.sample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import wang.relish.widget.vehicleedittext.VehicleKeyboardHelper;

/**
 * @author relish
 * @since 20200203
 */
public class MainFragment extends Fragment {

    public static MainFragment newInstance() {
        Bundle args = new Bundle();

        MainFragment fragment = new MainFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = LayoutInflater.from(container.getContext()).inflate(R.layout.activity_main, container, false);
        EditText vehicleEditView = v.findViewById(R.id.vet);
        VehicleKeyboardHelper.bind(vehicleEditView);
        return v;
    }
}


library/src/main/java/wang/relish/widget/vehicleedittext/VehicleEditText.java
package wang.relish.widget.vehicleedittext;

import android.app.Activity;
import android.content.Context;
import android.util.AttributeSet;

import androidx.appcompat.widget.AppCompatEditText;

/**
 * 绑定了车牌号键盘的输入框(EditText)
 *
 * @author Relish Wang
 * @since 2017/09/21
 */
public class VehicleEditText extends AppCompatEditText {

    public VehicleEditText(Context context) {
        super(context);
        init();
    }

    public VehicleEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public VehicleEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        VehicleKeyboardHelper.bind((Activity) getContext(),this, new VehicleKeyboardView(getContext()));
    }

    OnTouchListener mToucheListener;
    OnFocusChangeListener mFocusChangeListener;
    OnKeyListener mKeyListener;

    public void setOnTouchListener2(OnTouchListener l) {
        mToucheListener = l;
    }

    public void setOnFocusChangeListener2(OnFocusChangeListener l) {
        mFocusChangeListener = l;
    }

    public void setOnKeyListener2(OnKeyListener l) {
        mKeyListener = l;
    }
}


library/src/main/java/wang/relish/widget/vehicleedittext/VehicleKeyboardView.java
package wang.relish.widget.vehicleedittext;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.inputmethodservice.Keyboard;
import android.inputmethodservice.KeyboardView;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.WindowManager;

import androidx.core.content.ContextCompat;

import java.util.Arrays;
import java.util.List;

import relish.wang.vehicleedittext.R;

/**
 * 为KeyboardView编辑自定义样式
 *
 * @author Relish Wang
 * @since 2019/03/15
 */
/* package */ class VehicleKeyboardView extends KeyboardView {

    public static VehicleKeyboardView newInstance(Context context) {
        return new VehicleKeyboardView(context);
    }

    public VehicleKeyboardView(Context context) {
        super(context, null);
        init(context);
    }

    public VehicleKeyboardView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }


    public VehicleKeyboardView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    /**
     * 含省份的简称的输入键盘
     */
    private Keyboard mProvincesKeyBoard;
    /**
     * 含数字/字母的输入键盘
     */
    private Keyboard mLettersKeyBoard;

    private void init(Context context) {
        mProvincesKeyBoard = new Keyboard(context, R.xml.keyboard_car_number_provinces);
        mLettersKeyBoard = new Keyboard(context, R.xml.keyboard_car_number_letters);
        setKeyboard(mProvincesKeyBoard);
        setPreviewEnabled(false);// 无法在PopupWindow上展示PopupWindow，而预览恰恰就是一个PopupWindow
        if (context instanceof Activity) {
            WindowManager windowManager = ((Activity) context).getWindowManager();
            metrics = new DisplayMetrics();
            windowManager.getDefaultDisplay().getMetrics(metrics);
        }
    }

    void switchToProvinces() {
        if (getKeyboard() == mProvincesKeyBoard) return;
        setKeyboard(mProvincesKeyBoard);
    }

    void switchToLetters() {
        if (getKeyboard() == mLettersKeyBoard) return;
        setKeyboard(mLettersKeyBoard);
    }

    private Paint paint;
    private DisplayMetrics metrics;

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (paint == null) paint = new Paint();
        Keyboard keyboard = getKeyboard();
        if (keyboard == null) return;
        @SuppressLint("DrawAllocation") Rect bgRect = new Rect();
        getDrawingRect(bgRect);
        paint.setColor(ContextCompat.getColor(getContext(), R.color.keyboard_bg_color));
        canvas.drawRect(bgRect, paint);

        List<Keyboard.Key> keys = keyboard.getKeys();
        if (keys == null || keys.isEmpty()) return;
        paint.setTextAlign(Paint.Align.CENTER);
        Typeface font = Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL);
        paint.setTypeface(font);
        paint.setAntiAlias(true);
        for (Keyboard.Key key : keys) {
            drawCell(canvas, key);
        }
    }

    private void drawCell(Canvas canvas, Keyboard.Key key) {
        // 1 绘制背景（左右-3dp 上下-6dp）
        int _3dpToPx = (int) TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, 3, metrics);
        int _6dpToPx = (int) TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, 6, metrics);
        Drawable dr;
        if (Arrays.asList(
                Keyboard.KEYCODE_CANCEL,
                "ABC".hashCode(),
                "中文".hashCode(),
                Keyboard.KEYCODE_DELETE).contains(key.codes[0])) {
            dr = getContext().getResources().getDrawable(R.drawable.btn_gray);
        } else {
            dr = getContext().getResources().getDrawable(R.drawable.btn_white);
            if (key.codes[0] == -10086) {
                dr = getContext().getResources().getDrawable(R.drawable.btn_null);
            }
        }
        @SuppressLint("DrawAllocation")
        Rect bounds = new Rect(
                key.x + _3dpToPx,
                key.y + _6dpToPx,
                key.x + key.width - _3dpToPx,
                key.y + key.height - _6dpToPx);
        dr.setBounds(bounds);
        dr.draw(canvas);
        // 2 绘制文字
        if (key.label != null) {
            if (Arrays.asList("ABC".hashCode(), "中文".hashCode()).contains(key.codes[0])) {
                paint.setTextSize(bounds.height() - 5 * _6dpToPx);
            } else {
                paint.setTextSize(bounds.height() - 4 * _6dpToPx);
            }
            if (key.codes[0] == Keyboard.KEYCODE_DONE) {
                paint.setColor(getContext().getResources().getColor(R.color.black));
            } else {
                paint.setColor(getContext().getResources().getColor(R.color.black));
            }
            @SuppressLint("DrawAllocation")
            Rect rect = new Rect(
                    key.x,
                    key.y,
                    key.x + key.width,
                    key.y + key.height);
            Paint.FontMetricsInt fontMetrics = paint.getFontMetricsInt();
            int baseline =
                    (rect.bottom + rect.top - fontMetrics.bottom - fontMetrics.top) / 2;
            // 实现水平居中，drawText对应改为传入targetRect.centerX()
            paint.setTextAlign(Paint.Align.CENTER);
            canvas.drawText(key.label.toString(), rect.centerX(), baseline, paint);
        }
        // 3 绘制图标
        if (key.icon != null) {
            int _15dpToPx = 15 * _6dpToPx / 6;
            @SuppressLint("DrawAllocation") Rect iconRect = new Rect(
                    bounds.left + _15dpToPx,
                    bounds.top + _15dpToPx,
                    bounds.right - _15dpToPx,
                    bounds.bottom - _15dpToPx);
            key.icon.setBounds(iconRect);
            key.icon.draw(canvas);
        }
    }
}


app/src/main/java/wang/relish/widget/vehicleedittext/sample/WithFragmentActivity.java
package wang.relish.widget.vehicleedittext.sample;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

/**
 * @author relish
 * @since 20200203
 */
public class WithFragmentActivity extends AppCompatActivity {


    static void start(Context context){
        Intent intent = new Intent(context, WithFragmentActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_with_fragement);

        FragmentManager manager = getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        transaction.replace(R.id.fragment, MainFragment.newInstance());
        transaction.commit();
    }
}


library/src/main/java/wang/relish/widget/vehicleedittext/OnKeyboardActionAdapter.java
package wang.relish.widget.vehicleedittext;

import android.annotation.SuppressLint;
import android.inputmethodservice.Keyboard;
import android.inputmethodservice.KeyboardView;
import android.widget.EditText;

/**
 * 实现了键盘通用功能的OnKeyboardActionListener
 *
 * @author Relish Wang
 * @since 2017/09/18
 */
/* package */ abstract class OnKeyboardActionAdapter implements KeyboardView.OnKeyboardActionListener {

    private EditText mEditText;

    OnKeyboardActionAdapter(EditText mEditText) {
        this.mEditText = mEditText;
    }

    /**
     * （用于拓展）处理特殊的键盘按键响应
     *
     * @param primaryCode 主键值
     * @param keyCodes    所有键值
     */
    @SuppressLint("SetTextI18n")
    public boolean onKeyEvent(int primaryCode, int[] keyCodes) {
        return false;
    }

    /**
     * 关闭键盘
     */
    public abstract void close();

    @Override
    public void onKey(int primaryCode, int[] keyCodes) {
        if (onKeyEvent(primaryCode, keyCodes)) {
            return;
        }
        if (mEditText == null) return;
        if (primaryCode == Keyboard.KEYCODE_DELETE) { // 回退(backspace)
            int start = Math.max(mEditText.getSelectionStart(), 0);
            int end = Math.max(mEditText.getSelectionEnd(), 0);
            if (start != end) {
                mEditText.getText().delete(start, end);
            } else if (start > 0) {
                mEditText.getText().delete(start - 1, end);
            }
        } else if (primaryCode == Keyboard.KEYCODE_CANCEL) { // 完成（complete）
            close();
        } else if (primaryCode == Keyboard.EDGE_LEFT) {
            int start = mEditText.getSelectionStart();
            if (start > 0) mEditText.setSelection(start - 1);
        } else if (primaryCode == Keyboard.EDGE_RIGHT) {
            int start = mEditText.getSelectionStart();
            if (start < mEditText.length()) mEditText.setSelection(start + 1);
        } else {
            String text = Character.toString((char) primaryCode);
            int start = Math.max(mEditText.getSelectionStart(), 0);
            int end = Math.max(mEditText.getSelectionEnd(), 0);
            mEditText.getText().replace(Math.min(start, end), Math.max(start, end),
                    text, 0, text.length());
        }
    }

    @Override
    public void onPress(int primaryCode) {

    }

    @Override
    public void onRelease(int primaryCode) {

    }

    @Override
    public void onText(CharSequence text) {

    }

    @Override
    public void swipeLeft() {

    }

    @Override
    public void swipeRight() {

    }

    @Override
    public void swipeDown() {

    }

    @Override
    public void swipeUp() {

    }
}


library/src/main/java/wang/relish/widget/vehicleedittext/VehicleKeyboardHelper.java
package wang.relish.widget.vehicleedittext;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityManager;
import android.content.ComponentName;
import android.content.Context;
import android.inputmethodservice.Keyboard;
import android.os.IBinder;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.PopupWindow;

import androidx.annotation.Nullable;

import java.util.List;

import relish.wang.vehicleedittext.R;


/**
 * 输入框键盘绑定帮助类
 * <p>
 * 直接在布局中使用该控件:
 * <pre>
 * <code>
 * &lt;wang.relish.vehicle.VehicleEditText
 *      android:id="@+id/vet"
 *      android:layout_width="match_parent"
 *      android:layout_height="wrap_content" /&gt;
 * </code>
 * </pre>
 * <p>
 * 或者使用普通的EditText:
 *
 * <pre>
 * <code>
 *      EditText et = findViewById(R.id.et);
 *      VehicleKeyboardHelper.bind(et);
 * </code>
 * </pre>
 *
 * @author Relish Wang
 * @since 2017/09/18
 */
public class VehicleKeyboardHelper {

    private static final String PROVINCES =
            "京津渝沪冀晋辽吉黑苏浙皖闽赣鲁豫鄂湘粤琼川贵云陕甘青蒙桂宁新藏使领警学港澳";

    /**
     * 为EditText绑定车牌号输入键盘
     *
     * @param et 绑定的输入框控件(EditText)
     */
    public static void bind(EditText et) {
        bind(et, VehicleKeyboardView.newInstance(et.getContext()));
    }

    /**
     * 为EditText绑定车牌号输入键盘
     *
     * @param et       绑定的输入框控件(EditText)
     * @param keyboard 自定义键盘(KeyboardView)
     */
    public static void bind(EditText et, VehicleKeyboardView keyboard) {
        bind(et, keyboard, null);
    }

    /**
     * 为EditText绑定自定义键盘，并设置自定义键盘的特殊键的功能
     *
     * @param et       绑定的输入框控件(EditText)
     * @param keyboard 自定义键盘(KeyboardView)
     * @param l        键盘事件监听器
     */
    @SuppressWarnings("SameParameterValue")
    @SuppressLint("ClickableViewAccessibility")
    private static void bind(
            final EditText et,
            final VehicleKeyboardView keyboard,
            @Nullable VehicleKeyboardView.OnKeyboardActionListener l) {
        if (et == null) return;//throw new NullPointerException("EditText should not be NULL!");
        if (keyboard == null) return; // 表示使用系统键盘


        keyboard.setOnKeyboardActionListener(l != null
                ? l : new OnKeyboardActionAdapter(et) {
            @Override
            public void close() {
                hideCustomInput(et);
            }

            @Override
            public boolean onKeyEvent(int primaryCode, int[] keyCodes) {
                String s = et.getText().toString();
                if ("ABC".hashCode() == primaryCode) {
                    keyboard.switchToLetters();
                    return true;
                } else if ("中文".hashCode() == primaryCode) {
                    keyboard.switchToProvinces();
                    return true;
                } else {
                    // 除功能键以外的键
                    if (primaryCode != Keyboard.KEYCODE_DELETE && s.length() >= 8) {
                        return true;// 车牌号到最长长度了
                    }
                    if (PROVINCES.contains(String.valueOf((char) primaryCode))) {
                        if (s.length() == 0) {
                            keyboard.switchToLetters();
                        }
                        return false;
                    } else {
                        return super.onKeyEvent(primaryCode, keyCodes);
                    }
                }
            }
        });
        // 设置触摸的点击事件。为了显示光标。（直接设置InputType.TYPE_NULL的话光标会消失）
        et.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                int inType = et.getInputType();
                et.setInputType(InputType.TYPE_NULL);
                et.onTouchEvent(event);
                et.setInputType(inType);
                // 光标始终在内容的最后面
                et.setSelection(et.getText().length());
                if (et instanceof VehicleEditText) {
                    View.OnTouchListener toucheListener = ((VehicleEditText) et).mToucheListener;
                    if (toucheListener != null) {
                        return toucheListener.onTouch(v, event);
                    }
                    // 不拦截
                }
                return true;
            }
        });
        // 监听输入框的焦点事件：获得焦点显示自定义键盘；失去焦点收起自定义键盘
        et.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    hideSysInput(et);
                    showCustomInput(et, keyboard);
                } else {
                    hideCustomInput(et);
                }
                if (et instanceof VehicleEditText) {
                    View.OnFocusChangeListener changeListener =
                            ((VehicleEditText) et).mFocusChangeListener;
                    if (changeListener != null) {
                        changeListener.onFocusChange(v, hasFocus);
                    }
                }
            }
        });
        // 监听物理返回键，收起自定义键盘
        et.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    hideCustomInput(et);
                    return true;
                }
                if (et instanceof VehicleEditText) {
                    View.OnKeyListener keyListener = ((VehicleEditText) et).mKeyListener;
                    if (keyListener != null) {
                        return keyListener.onKey(v, keyCode, event);
                    }
                }
                return false;
            }
        });
        et.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 0) {
                    keyboard.switchToProvinces();// 当没有文字的时候键盘切换回省份
                }
            }
        });
    }

    /**
     * 展示自定义键盘
     *
     * @param et       输入框控件
     * @param keyboard 自定义键盘
     */
    private static void showCustomInput(final EditText et, final VehicleKeyboardView keyboard) {
        Object obj = et.getTag(R.id.keyboard);
        final PopupWindow keyboardWindow;
        if (obj == null) {
            keyboardWindow = new PopupWindow(
                    keyboard,
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT);
            keyboardWindow.setAnimationStyle(R.style.WindowAnimation);
            et.setTag(R.id.keyboard, keyboardWindow);
        } else {
            keyboardWindow = ((PopupWindow) obj);
        }
        if (keyboardWindow.isShowing()) return;
        keyboardWindow.setOutsideTouchable(false);
        keyboard.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    if (keyboardWindow != null && keyboardWindow.isShowing()) {
                        hideCustomInput(et);
                        return true;
                    }
                }
                return false;
            }
        });
        Window window = getWindow(et);
        if (window == null) return;
        View decorView = window.getDecorView();
        if (decorView == null) return;
        if (!isActivityRunning(et.getContext())) return;
        // 横竖屏切换时，Activity的onCreate可能还未执行完毕，还拿不到decorView
        decorView.post(() -> {
            // 解决底部被导航栏遮挡问题
            keyboardWindow.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
            keyboardWindow.showAtLocation(decorView, Gravity.BOTTOM, 0, 0);
            keyboardWindow.update();
        });
    }

    /**
     * 隐藏自定义键盘
     *
     * @param et 输入框控件
     */
    private static void hideCustomInput(EditText et) {
        if (et == null) return;
        if (!isActivityRunning(et.getContext())) return;
        et.clearFocus();
        Object tag = et.getTag(R.id.keyboard);
        if (tag == null) return;
        if (tag instanceof PopupWindow) {
            PopupWindow window = (PopupWindow) tag;
            if (window.isShowing()) {
                window.dismiss();
            }
        }
    }

    /**
     * 隐藏系统输入法
     *
     * @param et 输入框控件
     */
    private static void hideSysInput(EditText et) {
        IBinder windowToken = et.getWindowToken();
        if (windowToken != null) {
            InputMethodManager imm = (InputMethodManager)
                    et.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(windowToken, InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    private static Window getWindow(EditText et) {
        Context context = et.getContext();
        if (context instanceof Activity) {
            return ((Activity) context).getWindow();
        } else {
            Log.e("VehicleEditText", "EditText must have a Context which is a Activity.");
            return null;
        }
    }

    private static boolean isActivityRunning(Context context) {
        if (context == null) return false;
        if (context instanceof Activity) {
            return !((Activity) context).isFinishing();
        } else {
            return false;
        }
    }
}
